code/modules/client/preferences_ui.dm

```diff
@@ -0,0 +1,650 @@
+	)
+	.["key_bindings"] = list()
+	for(var/key in key_bindings)
+		for(var/kb_name in key_bindings[key])
+			.["key_bindings"][kb_name] += list(key)
+
+	// Get save slot name
+	.["save_slot_names"] = list()
+	if(!path)
+		return
+	var/savefile/S = new (path)
+	if(!S)
+		return
+	var/name
+	for(var/i in 1 to MAX_SAVE_SLOTS)
+		S.cd = "/character[i]"
+		S["real_name"] >> name
+		if(!name)
+			continue
+		.["save_slot_names"]["[i]"] = name
+
+
+
+/datum/preferences/ui_static_data(mob/user)
+	update_preview_icon()
+	. = list(
+		"all_species" = get_playable_species(),
+		"synth_types" = SYNTH_TYPES,
+		"bodytypes" = GLOB.body_types_list,
+		"ethnicities" = GLOB.ethnicities_list,
+		"citizenships" = CITIZENSHIP_CHOICES,
+		"religions" = RELIGION_CHOICES,
+		"corporate_relations" = CORP_RELATIONS,
+		"squads" = SELECTABLE_SQUADS,
+		"clothing" = list(
+			"underwear" = list(
+				"male" = GLOB.underwear_m,
+				"female" = GLOB.underwear_f,
+			),
+			"undershirt" = GLOB.undershirt_t,
+			"backpack" = GLOB.backpacklist,
+			"loadout" = GLOB.gear_datums
+		),
+		"hairstyles" = GLOB.hair_styles_list,
+		"facialhair" = GLOB.facial_hair_styles_list,
+		"hairgradient" = GLOB.hair_gradients_list,
+		"genders" = list(
+			"NEUTER",
+			"MALE",
+			"FEMALE",
+			"PLURAL",
+		),
+		"validation" = list(
+			"age" = list("min" = AGE_MIN, "max" = AGE_MAX),
+		),
+		"overflow_job" = SSjob?.overflow_role?.title,
+		"ui_styles" = UI_STYLES,
+	)
+	.["gearsets"] = list()
+	for(var/g in GLOB.gear_datums)
+		var/datum/gear/gearset = GLOB.gear_datums[g]
+		.["gearsets"][gearset.display_name] = list(
+			"name" = gearset.display_name,
+			"cost" = gearset.cost,
+			"slot" = gearset.slot,
+		)
+
+	.["jobs"] = list()
+	for(var/j in SSjob.joinable_occupations)
+		var/datum/job/job = j
+		var/rank = job.title
+		.["jobs"][rank] = list(
+			"color" = job.selection_color,
+			"description" = job.html_description,
+			"banned" = is_banned_from(user.ckey, rank),
+			"playtime_req" = job.required_playtime_remaining(user.client),
+			"account_age_req" = !job.player_old_enough(user.client),
+			"flags" = list(
+				"bold" = (job.job_flags & JOB_FLAG_BOLD_NAME_ON_SELECTION) ? TRUE : FALSE
+			)
+		)
+	.["special_occupations"] = list(
+		"Latejoin Xenomorph" = BE_ALIEN,
+		"Xenomorph when unrevivable" = BE_ALIEN_UNREVIVABLE,
+		"End of Round Deathmatch" = BE_DEATHMATCH,
+		"Prefer Squad over Role" = BE_SQUAD_STRICT
+	)
+
+	// Group keybinds by category
+	.["all_keybindings"] = list()
+	for(var/name in GLOB.keybindings_by_name)
+		var/datum/keybinding/kb = GLOB.keybindings_by_name[name]
+		.["all_keybindings"][kb.category] += list(list(
+			name = kb.name,
+			display_name = kb.full_name,
+			desc = kb.description,
+			category = kb.category,
+		))
+	.["mapRef"] = "player_pref_map"
+
+
+/datum/preferences/ui_act(action, list/params, datum/tgui/ui)
+	. = ..()
+	if(.)
+		return
+	var/client/current_client = CLIENT_FROM_VAR(usr)
+	var/mob/user = current_client.mob
```

Comment:
is this the best way to get user ?

Reply:
we don't know what proced preferences. Could be a ghost, could be lobby, could be whatever.
This is the safe proc to use